apps/mkcurtaindb/mkcurtaindb.8
.Dd April 16, 2024
.Dt mkcurtaindb 8
.Os mcplex.net
.Sh NAME
.Nm mkcurtaindb
.Nd create a mccurtain AS to IP dtabase from CAIDA routeviews files
.Sh SYNOPSIS
.Nm
.Op Fl o Ar dbfile
.Ar routeViewsIPv4File
.Ar routeViewsIPv6File
.Sh DESCRIPTION
.Nm
mqy be used to create a mccurtain AS to IP database from CAIDA
routeviews files.  This database is required by
.Xr mccurtaind 8
and
.Xr mccdb 1 .
.Pp

The command line arguments:
.Pp
.Bl -tag -width indent
.It Fl o Ar dbfile
Specify the paht to the output database file.  If this option is not
used, \fI/usr/local/etc/mccas2ip.db\fR will be assumed.
.It Ar routeViewsIPv4File
Specifies the path to the CAIDA routeviews file for IPv4.
.It Ar routeViewsIPv6File
Specifies the path the CAIDA routeviews file for IPv6.
.Sh EXAMPLES
All of the examples below assume that a CAIDA IPv4 routeviews file has
been downloaded and saved in \fIrv2.pfx2as.gz\fR and that a CAIDA IPv6
routeviews has been downloaded and saved in \fIrv6.pfx2as.gz\fR
.Pp
Create the database at the default path (\fI/usr/local/etc/mccas2ip.db\fR)
from the CAIDA IPv4 routeviews file in \fIrv2.pfx2as.gz\fR and the CAIDA
IPv6 routeviews file \fIrv6.pfx2as.gz\fR:
.Bd -literal
    # mkcurtaindb rv2.pfx2as.gz rv6.pfx2as.gz
.Ed
.Pp
Create the database \fI./mccas2ip.db\fR from the CAIDA IPv4 routeviews
file in \fIrv2.pfx2as.gz\fR and the CAIDA IPv6 routeviews file
\fIrv6.pfx2as.gz\fR, in uncompressed form.  This is the fastest format
to load, and hence is the preferred format for
.Xr mccdb 1 .  Despite being larger than the compressed format, it is
smaller than the size of the compressed CAIDA routeviews files:
.Bd -literal
    % mkcurtaindb -o ./mccas2ip.db rv2.pfx2as.gz rv6.pfx2as.gz
.Ed
.Pp
Create the database \fI./mccas2ip.db.bz2\fR from the CAIDA IPv4
routeviews file in \fIrv2.pfx2as.gz\fR and the CAIDA IPv6 routeviews
file \fIrv6.pfx2as.gz\fR, in bzip2 compressed form.  This format is
about half the size of the uncompressed format in most cases, at the
expense of a slower load time for
.Xr mccdb 1
and
.Xr mccurtaind 8 .  This is of litle consequence for
.Xr mccurtaind 8 since it only loads at startup and runs for long
periods (typically 24 hours to a week between reloads):
.Bd -literal
    % mkcurtaindb -o ./mccas2ip.db.bz2 rv2.pfx2as.gz rv6.pfx2as.gz
.Ed
.Pp
Create the database \fI./mccas2ip.db.gz\fR from the CAIDA IPv4
routeviews file in \fIrv2.pfx2as.gz\fR and the CAIDA IPv6 routeviews
file \fIrv6.pfx2as.gz\fR, in gzip compressed form.  This format is
about half the size of the uncompressed format in most cases, at the
expense of a slower load time for
.Xr mccdb 1 and
.Xr mccurtaind 8 .  This is of litle consequence for
.Xr mccurtaind 8 since it only loads at startup and runs for long
periods (typically 24 hours to a week between reloads):
.Bd -literal
    % mkcurtaindb -o ./mccas2ip.db.gz rv2.pfx2as.gz rv6.pfx2as.gz
.Ed
.Sh FILES
.Bl -tag -width indent
.It Pa /usr/local/etc/mccas2ip.db
The default mccurtain AS to IP database file.
.It Pa https://publicdata.caida.org/datasets/routing/routeviews-prefix2as/
The top level URL for the CAIDA IPv4 routeviews files.
.It Pa https://publicdata.caida.org/datasets/routing/routeviews6-prefix2as/
The top level URL for the CAIDA IPv6 routeviews files.
.El
.Sh SEE ALSO
.Xr mccdb 1 ,
.Xr mccurtain 1 ,
.Xr mccurtaind 8
.Sh AUTHORS
Copyright \(co 2026
.An Daniel W. McRobb Aq Mt dwm@mcplex.net


ChangeLog
2026-09-17  Daniel McRobb  <dwm@mbp22.mcplex.net>

	* tagged as mccurtain-0.0.38

	* apps/mccurtain/Makefile
	* apps/mccurtaind/Makefile
	* apps/mkcurtaindb/Makefile
	* classes/tests/Makefile
	Remove '-c' from compiler arguments when buulding dependencies.

2026-09-17  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.37

	* etc/mccurtaind.cfg.example
	Correct a comment.

2026-09-17  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as mccurtain-0.0.36

2026-09-17  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/Makefile
	* apps/mccurtaind/mccurtaind.cfg.5
	Added manpage for the mccurtaind configuration file.

2026-09-16  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* tagged as mccurtain-0.0.35

	* apps/mkcurtaindb/Makefile
	Correct tarprep target path for mccurtaindb.5

	* tagged as mccurtain-0.0.34

2026-09-16  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* apps/mccurtaind/DwmMcCurtainUdpServer.cc
	* apps/mccurtaind/DwmMcCurtainUdpServer.hh
	Check client access in UDP server.

	* etc/mccurtaind.cfg.example
	Add udpAddresses configuration.

	* apps/mkcurtaindb/mccurtaindb.5
	Add manpage for db file format.

	* apps/mccurtaind/mccurtaind.8
	* apps/mccurtaind/Makefile
	Add manpage for mccurtaind(8)

	* apps/mkcurtaindb/mkcurtaindb.8
	* apps/mkcurtaindb/Makefile
	Add manpage for mkcurtaindb(8)

	* tagged as mccurtain-0.0.33

	* classes/include/DwmMcCurtainResponses.hh
	include DwmIpv6Prefix.hh

2026-09-15  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.32

	* apps/mccurtaind/DwmMcCurtainServer.cc
	Populate IPv6 prefixes in responses to AS or country queries.

	* classes/include/DwmMcCurtainResponses.hh
	Add std::vector<Ipv6Prefix> to ASPrefixesResponse so that responses
	to queries for an AS or country can contain IPv6 prefixes.

	* apps/mccurtain/mccurtain.cc
	Added -4 and -6 command line options to select IPv4 or IPv6 prefix
	output when querying for an AS or country.  The default selects only
	IPv4 prefixes.

	* apps/mccurtain/mccurtain.cc
	* apps/mccurtaind/DwmMcCurtainResponder.cc
	* apps/mccurtaind/DwmMcCurtainResponder.hh
	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* classes/include/DwmMcCurtainRequests.hh
	* classes/include/DwmMcCurtainResponses.hh
	Added support for IPv6 address queries via our secure channel.

2026-09-15  Daniel McRobb  <dwm@mbp22.mcplex.net>

	* apps/mccdb/mccdb.cc
	Print country code when printing ASes within a country code.

2026-09-15  Daniel McRobb  <dwm@thrip.mcplex.net>

	* apps/mccdb/mccdb.1
	Added some examples.

	* apps/mccdb/mccdb.cc
	* apps/mccdb/mccdb.1
	Added '-4' and '-6' command line options to select only IPv4 or
	IPv6 prefixes.

2026-09-15  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* Renamed mccdbcat to mccdb.

	* apps/mccdbcat/mccdbcat.cc
	Allow selecting prefixes which match a given address, a given
	country or given AS.  Also allow display of ASes with a
	given country code.

2026-09-13  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.31

	* apps/mccdbcat/mccdbcat.cc
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/src/DwmMcCurtainDatabaseConfig.cc
	* etc/mccurtaind.cfg.example
	* scripts/updatecurtaindata.sh
	Be consistent about default db filename (now
	/usr/local/etc/mccas2ip.db).

	* tagged as mccurtain-0.0.30

	* apps/mccdbcat/Makefile
	* apps/mccdbcat/mccdbcat.1
	Added manpage for mccdbcat(1)

	* etc/mccurtaind.cfg.example
	Added to comment for dbFile.

	* apps/mccdbcat/mccdbcat.cc
	Use AS2IPDb::Load()

2026-09-13  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/include/DwmMcCurtainAS2IPDb.hh
	* classes/src/DwmMcCurtainAS2IPDb.cc
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/src/Makefile
	* classes/tests/Makefile
	* classes/tests/TestAS2IPDb.cc
	Added AS2IPDb class with static members to load and save
	AS2Ipv4Net and AS2Ipv6Net via common interface that automatically
	uses bz2, gz or no compression based on filename extension.

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	Allow loading AS to IP data from bzip2 or gzip files based on
	file extension.

	* apps/mkcurtaindb/mkcurtaindb.cc
	If output file name extension is ".bz2", use BZ2 compression.
	If output file name extension is ".gz", use gzip compression.
	Else use no compression.

2026-09-12  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmMcCurtainAS2Ipv4Net.hh
	* classes/src/DwmMcCurtainAS2Ipv4Net.cc
	* classes/tests/TestAS2Ipv4Net.cc
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/tests/TestAS2Ipv6Net.cc
	Save() and Load() will now use bzip2 compression if path extension
	is ".bz2", gzip compression if path extension is ".gz", and no
	compression in all other cases.

2026-09-12  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainDnsServer.hh
	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/include/DwmMcCurtainAS2Ipv4Net.hh
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/include/DwmMcCurtainIpv4Net2AS.hh
	* classes/include/DwmMcCurtainIpv6Net2AS.hh
	* classes/src/DwmMcCurtainAS2Ipv4Net.cc
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/src/DwmMcCurtainIpv4Net2AS.cc
	* classes/src/DwmMcCurtainIpv6Net2AS.cc
	* classes/tests/.gitignore
	* classes/tests/Makefile
	* classes/tests/TestAS2Ipv4Net.cc
	* classes/tests/TestAS2Ipv6Net.cc
	* classes/tests/TestCaidaV4Routeviews.cc
	* classes/tests/TestCaidaV6Routeviews.cc
	db file now only contains an AS2Ipv4Net and AS2Ipv6Net, since
	we can construct our Ipv4Net2AS and Ipv6Net2AS from them.  This
	reduces thr size of the uncompressed binaery bb file to a size
	less than the gzip'ed CAIDA routeviews files.

	I also added more I/O functionality to AS2Ipv4Net and AS2Ipv6Net

apps/mkcurtaindb/mkcurtaindb.cc
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2024, 2025, 2026
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file mkcurtain.cc
//!  \author Daniel W. McRobb
//!  \brief utility to create binary data file from routeviews data
//---------------------------------------------------------------------------

#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>

#include "DwmSysLogger.hh"
#include "DwmMcCurtainAS2IPDb.hh"
#include "DwmMcCurtainVersion.hh"

using namespace std;
namespace fs = std::filesystem;

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void Usage(const char *argv0)
{
  cerr << "Usage: " << argv0
       << " [-o outfile] routeViewsIPv4File routeViewsIPv6File\n";
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool PopulateV6Data(const string & caidarvfile,
                           Dwm::McCurtain::AS2Ipv6Net & as2ip)
{
  Dwm::McCurtain::CaidaV6Routeviews  rv;
  if (rv.Load(caidarvfile)) {
    rv.Aggregate();
    as2ip.Load(rv);
    return true;
  }
  else {
    cerr << "Failed to load data from '" << caidarvfile << "'\n";
  }
  return false;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool PopulateV4Data(const string & caidarvfile,
                           Dwm::McCurtain::AS2Ipv4Net & as2ip)
{
  Dwm::McCurtain::CaidaV4Routeviews  rv;
  if (rv.Load(caidarvfile)) {
    rv.Aggregate();
    as2ip.Load(rv);
    return true;
  }
  else {
    cerr << "Failed to load data from '" << caidarvfile << "'\n";
  }
  return false;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool SaveAS2IpData(const string & outfile,
                          Dwm::McCurtain::AS2Ipv4Net & asip4,
                          Dwm::McCurtain::AS2Ipv6Net & asip6)
{
  return Dwm::McCurtain::AS2IPDb::Save(outfile, asip4, asip6);
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  Dwm::SysLogger::Open("mkcurtaindb", LOG_PERROR|LOG_PID, "user");

  string  mccDbFile = "/usr/local/etc/mccas2ip.db";

  extern int  optind;
  int         optChar;
  while ((optChar = getopt(argc, argv, "o:")) != -1) {
    switch (optChar) {
      case 'o':
        mccDbFile = optarg;
        break;
      default:
        Usage(argv[0]);
        exit(1);
        break;
    }
  }

  if ((optind + 1) >= argc) {
    Usage(argv[0]);
    exit(1);
  }

  Dwm::McCurtain::AS2Ipv4Net  as2ip4;
  Dwm::McCurtain::AS2Ipv6Net  as2ip6;
  
  if (PopulateV4Data(argv[optind], as2ip4)) {
    if (PopulateV6Data(argv[optind+1], as2ip6)) {
      if (SaveAS2IpData(mccDbFile, as2ip4, as2ip6)) {
        return 0;
      }
      else {
        cerr << "Failed to save data to '" << mccDbFile << "'\n";
      }
    }
    else {
      cerr << "Failed to populate IPv6 data\n";
    }
  }
  else {
    cerr << "Failed to populate IPv4 data\n";
  }

  return 1;
}
